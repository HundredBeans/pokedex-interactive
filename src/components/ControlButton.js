import React, { useState } from "react";
import {
  Box,
  Button,
  Popover,
  PopoverTrigger,
  PopoverContent,
  PopoverBody,
  PopoverFooter,
  ButtonGroup,
  Image,
  FormControl,
  FormLabel,
  NumberInput,
  NumberInputField,
} from "@chakra-ui/react";
import control from "../assets/control.png";

function ControlButton(props) {
  const [isOpen, setIsOpen] = useState(false);
  const open = () => {
    setIsOpen(!isOpen);
  };
  const closePopover = () => {
    setIsOpen(false);
  };
  return (
    <React.Fragment>
      <Box
        bottom="0"
        left="50%"
        transform="translate(-50%, 0);"
        position="fixed"
        zIndex="banner"
        shadow="lg"
        rounded="full"
      >
        <Button
          variant="ghost"
          boxSize="100px"
          width="100%"
          rounded="full"
          onClick={open}
        >
          <Image src={control} />
        </Button>
      </Box>
      <Popover isOpen={isOpen} onClose={closePopover} closeOnBlur={false}>
        <PopoverTrigger>
          <Box bottom="250px" right="50%" position="fixed"></Box>
        </PopoverTrigger>
        <PopoverContent width="50%" mx="auto" shadow="dark-lg">
          <PaginationForm {...props} />
        </PopoverContent>
      </Popover>
    </React.Fragment>
  );
}

const PaginationForm = (props) => {
  return (
    <React.Fragment>
      <PopoverBody>
        <FormControl>
          <FormLabel htmlFor="nickname" textTransform="capitalize">
            Change Page
          </FormLabel>
          <NumberInput
            textAlign="center"
            size="md"
            max={props.maxPage}
            value={props.page}
            min={1}
            onChange={(e) => props.onChangePage(e)}
          >
            <NumberInputField />
          </NumberInput>
        </FormControl>
      </PopoverBody>
      <PopoverFooter textAlign="center">
        <ButtonGroup size="sm">
          <Button
            colorScheme="teal"
            disabled={props.page === 1}
            onClick={props.prevPage}
          >
            Back
          </Button>
          <Button
            colorScheme="teal"
            disabled={props.page === props.maxPage}
            onClick={props.nextPage}
          >
            Next
          </Button>
        </ButtonGroup>
      </PopoverFooter>
    </React.Fragment>
  );
};

export default ControlButton;
