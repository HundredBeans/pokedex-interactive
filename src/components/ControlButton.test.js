import { cleanup, fireEvent, waitFor } from "@testing-library/react";
import React from "react";
import { render } from "../test-utils";
import ControlButton from "./ControlButton";

describe("ControlButton.js test", () => {
  afterEach(() => {
    cleanup();
  });

  it("renders the component", () => {
    const renderedComponent = render(<ControlButton />);
    expect(renderedComponent).toMatchSnapshot();
  });

  it("can open and close the popOver", async () => {
    const renderedComponent = render(<ControlButton />);
    // Check if the popover is not open
    const popOverContent = renderedComponent.getByText("Change Page");
    expect(popOverContent).not.toBeVisible();
    const popoverTriggerBtn = renderedComponent.getByRole("button");
    fireEvent.click(popoverTriggerBtn);
    await waitFor(() => {
      // Check if popover visible
      expect(popOverContent).toBeVisible();
    });
    fireEvent.click(popoverTriggerBtn);
    await waitFor(() => {
      // Check if popover is closed
      expect(popOverContent).not.toBeVisible();
    });
  });

  test("pagination form is working", async () => {
    const prevPageMock = jest.fn();
    const nextPageMock = jest.fn();
    const onChangePageMock = jest.fn();
    const renderedComponent = render(
      <ControlButton
        page={1}
        maxPage={10}
        onChangePage={onChangePageMock}
        prevPage={prevPageMock}
        nextPage={nextPageMock}
      />
    );
    // Check if the popover is not open
    const popOverContent = renderedComponent.getByText("Change Page");
    const popoverTriggerBtn = renderedComponent.getByRole("button");
    fireEvent.click(popoverTriggerBtn);
    // Wait for the popover is visible
    await waitFor(() => {
      // Check if popover visible
      expect(popOverContent).toBeVisible();
    });
    const prevPageBtn = renderedComponent.getByText("Back");
    expect(prevPageBtn).toBeDisabled();
    const nextPageBtn = renderedComponent.getByText("Next");
    expect(nextPageBtn).not.toBeDisabled();
    fireEvent.click(nextPageBtn);
    expect(nextPageMock).toBeCalled();
    // Test Pagination Input
    const paginationInput = renderedComponent.getByLabelText("Change Page");
    fireEvent.change(paginationInput, {
      target: {
        value: 10,
      },
    });
    await waitFor(() => {
      expect(onChangePageMock).toBeCalled();
    });
  });
});
