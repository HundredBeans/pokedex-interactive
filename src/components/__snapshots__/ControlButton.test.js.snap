// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlButton.js test renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; top: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; right: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; bottom: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; right: 0px;"
      />
    </div>
    <div>
      <div
        class="css-1ue797s"
      >
        <button
          class="chakra-button css-17i1qg7"
          type="button"
        >
          <img
            class="chakra-image__placeholder css-0"
          />
        </button>
      </div>
      <div
        aria-controls="popover-content-3"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="css-17xwwk7"
        id="popover-trigger-3"
      />
      <div
        class="css-11r8ny"
        data-popper-escaped=""
        data-popper-placement="bottom"
        data-popper-reference-hidden=""
        style="--popper-arrow-shadow-color: #E2E8F0; visibility: hidden; margin: 0px; --popper-transform-origin: top center; position: absolute; left: 0px; top: 0px; transform: translate(0px, 8px);"
      >
        <section
          aria-describedby="popover-body-3"
          class="chakra-popover__content css-17rezgt"
          id="popover-content-3"
          role="dialog"
          style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
          tabindex="-1"
        >
          <div
            class="chakra-popover__body css-1ews2c8"
            id="popover-body-3"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1wjdhu5"
                for="nickname"
                id="field-1-label"
              >
                Change Page
              </label>
              <div
                class="css-16krjww"
              >
                <input
                  aria-disabled="false"
                  aria-valuemax="9007199254740991"
                  aria-valuemin="1"
                  autocomplete="off"
                  autocorrect="off"
                  class="css-xvw9gz"
                  id="nickname"
                  inputmode="decimal"
                  pattern="[0-9]*(.[0-9]+)?"
                  role="spinbutton"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <footer
            class="chakra-popover__footer css-vxuicp"
          >
            <div
              class="chakra-button__group css-1ezzth2"
              role="group"
            >
              <button
                class="chakra-button css-1k6fjms"
                type="button"
              >
                Back
              </button>
              <button
                class="chakra-button css-1k6fjms"
                disabled=""
                type="button"
              >
                Next
              </button>
            </div>
          </footer>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-1ue797s"
    >
      <button
        class="chakra-button css-17i1qg7"
        type="button"
      >
        <img
          class="chakra-image__placeholder css-0"
        />
      </button>
    </div>
    <div
      aria-controls="popover-content-3"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="css-17xwwk7"
      id="popover-trigger-3"
    />
    <div
      class="css-11r8ny"
      data-popper-escaped=""
      data-popper-placement="bottom"
      data-popper-reference-hidden=""
      style="--popper-arrow-shadow-color: #E2E8F0; visibility: hidden; margin: 0px; --popper-transform-origin: top center; position: absolute; left: 0px; top: 0px; transform: translate(0px, 8px);"
    >
      <section
        aria-describedby="popover-body-3"
        class="chakra-popover__content css-17rezgt"
        id="popover-content-3"
        role="dialog"
        style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
        tabindex="-1"
      >
        <div
          class="chakra-popover__body css-1ews2c8"
          id="popover-body-3"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1wjdhu5"
              for="nickname"
              id="field-1-label"
            >
              Change Page
            </label>
            <div
              class="css-16krjww"
            >
              <input
                aria-disabled="false"
                aria-valuemax="9007199254740991"
                aria-valuemin="1"
                autocomplete="off"
                autocorrect="off"
                class="css-xvw9gz"
                id="nickname"
                inputmode="decimal"
                pattern="[0-9]*(.[0-9]+)?"
                role="spinbutton"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <footer
          class="chakra-popover__footer css-vxuicp"
        >
          <div
            class="chakra-button__group css-1ezzth2"
            role="group"
          >
            <button
              class="chakra-button css-1k6fjms"
              type="button"
            >
              Back
            </button>
            <button
              class="chakra-button css-1k6fjms"
              disabled=""
              type="button"
            >
              Next
            </button>
          </div>
        </footer>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
